Add an article version of my WWDC talk on how to write macros (#2408)
apple/swift-syntax: Sources/SwiftSyntaxMacros/SwiftSyntaxMacros.docc/SwiftSyntaxMacros.md

```

`@freestanding(expression)` declares the macro with the freestanding expression macro role. This means that the macro can be used wherever an expression can be used and that its uses is indicated by the hash character, like `#stringify`. More macro roles are discussed below in [Macro Roles](#Macro-Roles).

The next part of the macro declaration looks like a function signature. It takes a generic argument of type `T` and outputs a tuple containing the result, again of type `T`, as well as a string containing the source code of the macro’s argument. All arguments need to type check and the macro’s result type is inferred from the return type. If the arguments don’t type check or the result type doesn’t match its surrounding context, the compiler will emit an error without applying the macro expansion. This is different to, for example, C macros, which are evaluated at the pre-processor stage before type-checking.

The last part of the macro links the macro declaration to the macro implementation. `#externalMacro` tells the compiler that to expand the macro, it needs to use the `StringifyMacro` type in the `MyMacroMacros` module.

### Macro Implementation

> The `StringifyMacro` type is declared in a separate target, which is a compiler plug-in. The compiler will send the source code of the entire macro expression to that plug-in. The first thing that the macro plug-in does, is to parse the source code of the macro into a SwiftSyntax tree. This tree is a source-accurate, structural representation of the macro, and it will be the basis on which the macro operates. 

Author: Changed to 

> The compiler will send the macro expression to that plug-in. The plugin processes the macro expression as a SwiftSyntax tree, which is a source-accurate, structural representation of the macro, and it will be the basis on which the macro operates.

    case .actor: return .actor
    case .class: return .class
    ...
    }
  }
}
```

### Declare EnumSubset

Since both the initializer and the computed property are members of `TypeDeclarationKeyword`, the `EnumSubset` macro needs to be an attached member macro. In the following, we will only generate the initializer. Generating the computed property is analogous.

The macro declaration looks as follows.

```swift
@attached(member, names: named(init))
public macro EnumSubset<Superset>() = #externalMacro(module: "MyMacroMacros", type: "EnumSubsetMacro")
```

Compared to the `stringify` macro, `EnumSubset` differs in two ways:
1. It is declared as an attached member macro and defines the names of the members it introduces. Declaring the introduced names improves the compiler’s performance: When accessing a member that is not declared in the macro’s names, the compiler doesn’t need to try expanding the macro to find it.
2. `EnumSubset` doesn’t take any argument. Instead, it defines a generic parameter that is used to customize the superset – in the example this is `Keyword`. Using a generic parameter ensures that the superset type exists.

### Implement EnumSubset

> A possible implementation of `EnumSubsetMacro` can look as follows. 
> In the following sections, we explore techniques for how to write this code.

Reviewer: ```suggestion
A possible implementation of `EnumSubsetMacro` is as follows.
```

The latter part isn't needed IMO, you explicitly mention the sections in comments anyway.

```swift
public enum EnumSubsetMacro: MemberMacro {
  public static func expansion(
    of attribute: AttributeSyntax,
    providingMembersOf declaration: some DeclGroupSyntax,
    in context: some MacroExpansionContext
  ) throws -> [DeclSyntax] {
    guard let enumDecl = declaration.as(EnumDeclSyntax.self) else {
      // Emit an error here, see section *Emitting Errors*
      return []
    }

    // Extract the name of the generic parameter. 

>     // See section *Inspect the SwiftSyntax Tree* how to build this expresion.

Reviewer: ```suggestion
    // See section *Inspect the SwiftSyntax Tree* for more details on building this expression.
```

    guard let supersetType = attribute,
      .attributeName.as(SimpleTypeIdentifierSyntax.self)?
      .genericArgumentClause?
      .arguments.first?
      .argumentType else {

>       // Emit an error here.

Reviewer: I think it would be good to include emitting the error here to serve as a complete example.

Author: Added the error handling to the example.

      return []
    }
    
    // Extract all the enum elements
    let members = enumDecl.memberBlock.members
    let caseDecls = members.compactMap { $0.decl.as(EnumCaseDeclSyntax.self) }
    let elements = caseDecls.flatMap { $0.elements }
    
    // Build the initializer using a result builder
    let initializer = try InitializerDeclSyntax("init?(_ superset: \(supersetType))") {
      try SwitchExprSyntax("switch superset") {
        for element in elements {
          SwitchCaseSyntax(
            """
            case .\(element.identifier):
              self = .\(element.identifier)
            """
          )
        }
        SwitchCaseSyntax("default: return nil")
      }
    }
    
    return [DeclSyntax(initializer)]
  }
}
```

## Inspect the SwiftSyntax Tree

Navigating the syntax tree is not trivial. When the syntax tree seems overly complicated, this is usually because it needs to represent the entire Swift language. For example, each enum case declaration can declare multiple elements because it is possible to declare multiple enum elements after a single case keyword like:
```swift
case `actor`, `class`

> ```

Reviewer: I know why you have this here, but it's not needed IMO. I'd just start with:
> One of the best ways to explore the syntax tree is by example....

Usually, the best way to explore the syntax trees is by example. <https://swift-ast-explorer.com> is a great webpage that allows the shows the syntax tree of entered Swift code.  

> > Experiment: Try pasting the following code to <https://swift-ast-explorer.com> and follow the syntax tree to understand the above code that extracts the name of the generic parameter in `EnumSubsetMacro.expansion`.

Reviewer: ```suggestion
> Experiment: Try pasting the following code into <https://swift-ast-explorer.com> and compare the generated syntax tree to the above code that extracts the name of the generic parameter in `EnumSubsetMacro.expansion`.
```

> ```
> @EnumSubset<Keyword>
> ```

Alternatively, the syntax tree can be printed in the debugger. When the debugger is stopped and `node` is a syntax node, the syntax tree can be printed using `po node`. This produces a syntax tree as shown in the [Macro Implementation](#Macro-Implementation) section.

## Building Syntax Nodes

There are three core approaches to build syntax nodes: Parsing from string literals, result builder initializers and memberwise initializers. 

> The `stringify` macro created its result type, an `ExprSyntax` from a string literal. As described in that section, this invokes the Swift parser to parse the contents of the string literal into a syntax tree. This technique is great for statically known trees or trees with a fixed number of parameters.

Reviewer: ```suggestion
The `stringify` macro created its result type, an `ExprSyntax`, from a string literal. As described in that section, this invokes the Swift parser to parse the contents of the string literal into a syntax tree. This technique works well for statically known trees, or trees with a fixed number of parameters.
```

> The `EnumSubset` macro uses a result builder initializer to generate the `InitializerDeclSyntax` and `SwitchExprSyntax`. The initializer takes a header – the `swiftch` keyword and subject – and a trailing closure, which is a result builder that adds a `SwitchCaseSyntax` for every `case` item in the `switch` statement. Result builder initializers are a great tool for repetitive syntax constructs, like the `switch` in `EnumSubset`. Also note how `EnumSubsetMacro` combines result builders and string literals. It uses the result builder to generate the `SwitchExprSyntax` but the cases are constructed using string interpolation.

Reviewer: ```suggestion
The `EnumSubset` macro uses a result builder initializer to generate the `InitializerDeclSyntax` and `SwitchExprSyntax`. The initializer takes a header – the `switch` keyword and subject – and a trailing closure, which is a result builder that adds a `SwitchCaseSyntax` for every `case` item in the `switch` statement. Result builder initializers are a great tool for repetitive syntax constructs, like the `switch` in `EnumSubset`. Also note how `EnumSubsetMacro` combines result builders and string literals. It uses the result builder to generate the `SwitchExprSyntax` but the cases are constructed using string interpolation.
```

> Finally, every syntax type has a memberwise initializer with which a syntax node can be created by specifying all its children. This is the most low-level way of creating syntax nodes and has the most flexibility. It is also a good option for nodes that are not representable by string literals and don’t have result builder initializers.

Reviewer: ```suggestion
Finally, every syntax type has a memberwise initializer with which a syntax node can be created by specifying all its children.
```

Or reword the last two sentences - "good option" is a bit odd since it's actually the *only* option. Maybe:
> Finally, every syntax type has a memberwise initializer with which a syntax node can be created by specifying all its children. This initializer gives you full control when creating a syntax node, but is also the most verbose.

## Emitting Errors

Macros are not always applicable. For example, it doesn’t make sense to apply the `EnumSubset` macro on a type that is not an enum. If a macro is used in ways that it doesn’t support, it is always better to emit custom error messages that tells the adopter about what’s going wrong, instead of having them read the generated code to debug the macro.

Macro errors can be represented by any type that conforms to the Swift `Error` protocol. For example, an enum with a single case can be used to describe the error message if `EnumSubset` is applied to a type that's not an enum.

```swift
enum EnumSubsetError: CustomStringConvertible, Error {
  case onlyApplicableToEnum
  
  var description: String {
    switch self {
    case .onlyApplicableToEnum: return "@EnumSubset can only be applied to an enum"
    }
  }
}
```

An error can be emitted by throwing it from the expansion function. Its message will be shown inline with any other compilation errors.

> The ``MacroExpansionContext/addDiagnostics(from:node:)`` method on the ``MacroExpansionContext`` allows further customization of the error message, like showing the error message at a different location than the attribute, generating warnings, or even providing Fix-Its in Xcode.

Reviewer: > There's two ways to emit this error:
>   1. Throw the error from the expansion function. This will generate a compilation error on the line of the attribute.
>   2. Add it as a diagnostic with  ``MacroExpansionContext/addDiagnostics(from:node:)``. This allows further customization, eg. changing the location, emitting a warning instead of an error, or even providing Fix-Its.

Author: Changed to

> There are two ways to emit this error:
>
> 1. Throw an error from the `expansion` function. This generates a compilation error on the line of the attribute. This is the technique that the above example uses. 
> 2. Add it as a diagnostic with  ``MacroExpansionContext/diagnose(_:)``. This allows further customization, like changing the location, emitting a warning instead of an error, or even providing Fix-Its.
